bson/__init__.py

```diff
@@ -1166,6 +1166,9 @@ def decode_file_iter(file_obj, codec_options=DEFAULT_CODEC_OPTIONS):
         elif len(size_data) != 4:
             raise InvalidBSON("cut off in middle of objsize")
         obj_size = _UNPACK_INT_FROM(size_data, 0)[0] - 4
+        if obj_size < 0:
+            raise InvalidBSON(
+                "invalid objsize: {!r} -> {}".format(size_data, obj_size))
```

Comment:
What if instead of raising InvalidBSON explicitly, we let _bson_to_dict handle the error?:
```python
if obj_size <= 0:
    # Invalid object size, continue and let _bson_to_dict raise an error.
    elements = size_data
else:
    elements = size_data + file_obj.read(obj_size)
yield _bson_to_dict(elements, codec_options)
```

Then the following two files would raise the error from the same place (`_bson_to_dict`):
```python
import bson
with open('example1.bson', 'wb') as outfile:
  outfile.write(b'\x00\x00\x00\x00')
with open('example1.bson', 'rb') as infile:
    list(bson.decode_file_iter(infile))
```
```python
import bson
with open('example2.bson', 'wb') as outfile:
  outfile.write(b'\x05\x00\x00\x00')
with open('example2.bson', 'rb') as infile:
    list(bson.decode_file_iter(infile))

```

Reply:
I like this idea. I'm using min() here since `read(0)` will do what it Should™ (that is, read zero bytes and return `b''`).